tests/src/test/scala/akka/kafka/scaladsl/RebalanceExtSpec.scala

```diff
@@ -0,0 +1,427 @@
+    }
+  }
+
+  def consumerSettings(group: String,
+                       maxPollRecords: String,
+                       partitionAssignmentStrategy: String): ConsumerSettings[String, String] =
+    consumerDefaults
+      .withGroupId(group)
+      .withCloseTimeout(5.seconds)
+      .withPollInterval(300.millis)
+      .withPollTimeout(200.millis)
+      .withProperty(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, maxPollRecords) // 500 is the default value
+      .withProperty(ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG, partitionAssignmentStrategy)
+
+  def subscribeAndConsumeMessages(
+      clientId: String,
+      perPartitionMessageCount: Int,
+      topicPartitionFutureMap: Map[String, Promise[Done]],
+      messageStoreAndAck: Map[Int, (AtomicInteger, Promise[Done], Promise[Done])],
+      consumerSettings: ConsumerSettings[String, String],
+      subscription: AutoSubscription,
+      tpCount: Int,
+      ptCount: Int,
+      sharedKillSwitch: SharedKillSwitch,
+      businessFlow: (String,
+                     Int,
+                     Map[String, Promise[Done]],
+                     Map[Int, (AtomicInteger, Promise[Done], Promise[Done])],
+                     SharedKillSwitch,
+                     String) => Flow[CommittableMessage[String, String], CommittableOffset, NotUsed]
+  ): (Consumer.Control, Future[Seq[Future[Done]]]) = {
+    val logPrefix = clientId
+    val rebalanceListener = system.actorOf(Props(new RebalanceListenerActor(clientId, log)))
+    Consumer
+      .committablePartitionedSource(
+        consumerSettings.withClientId(clientId),
+        subscription
+          .withPartitionAssignmentHandler(assignmentHandler(clientId))
+          .withRebalanceListener(rebalanceListener)
+      )
+      .map {
+        case (topicPartition, topicPartitionStream) =>
+          log.debug(s"$logPrefix::Consuming partitioned source clientId: $clientId, for tp: $topicPartition")
+          val innerStream: Source[ConsumerMessage.CommittableOffsetBatch, NotUsed] = topicPartitionStream
+            .via(sharedKillSwitch.flow)
+            .via(
+              businessFlow(
+                clientId,
+                perPartitionMessageCount,
+                topicPartitionFutureMap,
+                messageStoreAndAck,
+                sharedKillSwitch,
+                logPrefix
+              )
+            )
+            .via(Committer.batchFlow(committerDefaults.withMaxBatch(1)))
+          innerStream.runWith(Sink.ignore)
+      }
+      .via(sharedKillSwitch.flow)
+      .toMat(Sink.seq)(Keep.both)
+      .run()
+  }
+
+  // create topic-partition map
+  def createTopicMapsAndPublishMessages(
+      topicCount: Int,
+      partitionCount: Int,
+      perPartitionMessageCount: Int
+  ): (Map[String, MSet[TopicPartition]],
+      Map[String, Promise[Done]],
+      Map[Int, String],
+      Set[String],
+      Map[Int, String],
+      Seq[Future[Done]],
+      Map[Int, (AtomicInteger, Promise[Done], Promise[Done])]) = {
```

Comment:
This return signature definitely deserves its own type!

Reply:
Introducing more types will reduce the herds of parameters passed to other methods too.

Reply:
done